docs/proposal/circuit_breaker.md

```diff
@@ -0,0 +1,374 @@
+
+<div align="center">
+    <img src="./pics/circuit_breaker_example.png" />
+</div>
+
+Here is the example based on the figure above:
+
+1. When the requests from the frontend service to the target service `forecast` do not exceed the configured maximum connection count, they are allowed through.
+2. When the requests from the frontend service to the target service `forecast` do not exceed the configured maximum pending requests count, they enter the connection pool to wait.
+3. When the requests from the frontend service to the target service `forecast` exceed the configured maximum pending requests count, they are directly rejected.
+
+Take threshold `max_connection,` for example; if the number of active connections exceeds the threshold, it will open a circuit breaker.
+
+`canCreateConnection` simply checks whether the number of active actions is below the cluster or per host's thresholds.
+
+```c++
+bool canCreateConnection(Upstream::ResourcePriority priority) const override {
+    if (stats().cx_active_.value() >= cluster().resourceManager(priority).maxConnectionsPerHost()) {
+        return false;
+    }
+    return cluster().resourceManager(priority).connections().canCreate();
+}
+```
+
+If cannot create a new connection, the `upstream_cx_overflow_` counter in cluster traffic stats will be increased.
+
+```c++
+ConnPoolImplBase::tryCreateNewConnection(float global_preconnect_ratio) {
+    const bool can_create_connection = host_->canCreateConnection(priority_);
+
+    if (!can_create_connection) {
+        host_->cluster().trafficStats()->upstream_cx_overflow_.inc();
+    }
+
+    // If we are at the connection circuit-breaker limit due to other upstreams having
+    // too many open connections, and this upstream has no connections, always create one, to
+    // prevent pending streams being queued to this upstream with no way to be processed.
+    if (can_create_connection || (ready_clients_.empty() && busy_clients_.empty() &&
+                                    connecting_clients_.empty() && early_data_clients_.empty())) {
+        ENVOY_LOG(debug, "creating a new connection (connecting={})", connecting_clients_.size());
+        // here are some logics for establishing a connection 
+    } else {
+        ENVOY_LOG(trace, "not creating a new connection: connection constrained");
+        return ConnectionResult::NoConnectionRateLimited;
+    }
+}
+```
+
+Envoy also supports outlier detection. If an endpoint produces too many exceptions (e.g., returning 5xx HTTP status), it will be temporarily removed from the connection pool.
+
+<div align="center">
+    <img src="./pics/outlier_detection.png" />
+</div>
+
+After some time, it will be added back, but if it continues to fail, it will be removed again. Each time it's removed, the wait time to be added back increases.
```

Comment:
```suggestion
After some time, it will be added back, but if it continues to fail, it will be removed again. The wait time for re-adding increases with each removal.
```